File: mojo/public/cpp/bindings/lib/control_message_handler.cc
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "mojo/public/cpp/bindings/lib/control_message_handler.h"

#include <stddef.h>
#include <stdint.h>
#include <utility>

#include "base/logging.h"
#include "mojo/public/cpp/bindings/lib/message_builder.h"
#include "mojo/public/cpp/bindings/lib/serialization.h"
#include "mojo/public/interfaces/bindings/interface_control_messages.mojom.h"

namespace mojo {
namespace internal {

// static
bool ControlMessageHandler::IsControlMessage(const Message* message) {
  return message->header()->name == interface_control::kRunMessageId ||
         message->header()->name == interface_control::kRunOrClosePipeMessageId;
}

ControlMessageHandler::ControlMessageHandler(uint32_t interface_version)
    : interface_version_(interface_version) {
}

ControlMessageHandler::~ControlMessageHandler() {
}

bool ControlMessageHandler::Accept(Message* message) {
  if (message->header()->name == interface_control::kRunOrClosePipeMessageId)
    return RunOrClosePipe(message);

  NOTREACHED();
  return false;
}

bool ControlMessageHandler::AcceptWithResponder(
    Message* message,
    MessageReceiverWithStatus* responder) {
  if (message->header()->name == interface_control::kRunMessageId)
    return Run(message, responder);

  NOTREACHED();
  return false;
}

bool ControlMessageHandler::Run(Message* message,
                                MessageReceiverWithStatus* responder) {
  interface_control::internal::RunMessageParams_Data* params =
      reinterpret_cast<interface_control::internal::RunMessageParams_Data*>(
          message->mutable_payload());
  interface_control::RunMessageParamsPtr params_ptr;
  Deserialize<interface_control::RunMessageParamsDataView>(params, &params_ptr,
                                                           &context_);
  auto& input = *params_ptr->input;
  interface_control::RunOutputPtr output = interface_control::RunOutput::New();
  if (input.is_query_version()) {
    output->set_query_version_result(
        interface_control::QueryVersionResult::New());
    output->get_query_version_result()->version = interface_version_;
  } else {
    output.reset();
  }

  auto response_params_ptr = interface_control::RunResponseMessageParams::New();
  response_params_ptr->output = std::move(output);
  size_t size =
      PrepareToSerialize<interface_control::RunResponseMessageParamsDataView>(
          response_params_ptr, &context_);
  ResponseMessageBuilder builder(interface_control::kRunMessageId, size,
                                 message->request_id());

  interface_control::internal::RunResponseMessageParams_Data* response_params =
      nullptr;
  Serialize<interface_control::RunResponseMessageParamsDataView>(
      response_params_ptr, builder.buffer(), &response_params, &context_);
  bool ok = responder->Accept(builder.message());
  ALLOW_UNUSED_LOCAL(ok);
  delete responder;

  return true;
}

bool ControlMessageHandler::RunOrClosePipe(Message* message) {
  interface_control::internal::RunOrClosePipeMessageParams_Data* params =
      reinterpret_cast<
          interface_control::internal::RunOrClosePipeMessageParams_Data*>(
          message->mutable_payload());
  interface_control::RunOrClosePipeMessageParamsPtr params_ptr;
  Deserialize<interface_control::RunOrClosePipeMessageParamsDataView>(
      params, &params_ptr, &context_);
  auto& input = *params_ptr->input;
  if (input.is_require_version())
    return interface_version_ >= input.get_require_version()->version;

  return false;
}

}  // namespace internal
}  // namespace mojo

File: mojo/public/js/lib/control_message_handler.js
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

define("mojo/public/js/lib/control_message_handler", [
  "mojo/public/js/codec",
  "mojo/public/interfaces/bindings/interface_control_messages.mojom",
  "mojo/public/js/validator",
], function(codec, controlMessages, validator) {

  var Validator = validator.Validator;

  function validateControlRequestWithResponse(message) {
    var messageValidator = new Validator(message);
    var error = messageValidator.validateMessageIsRequestExpectingResponse();
    if (error !== validator.validationError.NONE) {
      throw error;
    }

    if (message.getName() != controlMessages.kRunMessageId) {
      throw new Error("Control message name is not kRunMessageId");
    }

    // Validate payload.
    error = controlMessages.RunMessageParams.validate(messageValidator,
        message.getHeaderNumBytes());
    if (error != validator.validationError.NONE) {
      throw error;
    }
  }

  function validateControlRequestWithoutResponse(message) {
    var messageValidator = new Validator(message);
    var error = messageValidator.validateMessageIsRequestWithoutResponse();
    if (error != validator.validationError.NONE) {
      throw error;
    }

    if (message.getName() != controlMessages.kRunOrClosePipeMessageId) {
      throw new Error("Control message name is not kRunOrClosePipeMessageId");
    }

    // Validate payload.
    error = controlMessages.RunOrClosePipeMessageParams.validate(
        messageValidator, message.getHeaderNumBytes());
    if (error != validator.validationError.NONE) {
      throw error;
    }
  }

  function runOrClosePipe(message, interface_version) {
    var reader = new codec.MessageReader(message);
    var runOrClosePipeMessageParams = reader.decodeStruct(
        controlMessages.RunOrClosePipeMessageParams);
    return interface_version >=
        runOrClosePipeMessageParams.input.require_version.version;
  }

  function run(message, responder, interface_version) {
    var reader = new codec.MessageReader(message);
    var runMessageParams =
        reader.decodeStruct(controlMessages.RunMessageParams);
    var runOutput = null;

    if (runMessageParams.input.query_version) {
      runOutput = new controlMessages.RunOutput();
      runOutput.query_version_result = new
          controlMessages.QueryVersionResult({'version': interface_version});
    }

    var runResponseMessageParams = new
        controlMessages.RunResponseMessageParams();
    runResponseMessageParams.output = runOutput;

    var messageName = controlMessages.kRunMessageId;
    var payloadSize = controlMessages.RunResponseMessageParams.encodedSize;
    var requestID = reader.requestID;
    var builder = new codec.MessageWithRequestIDBuilder(messageName,
        payloadSize, codec.kMessageIsResponse, requestID);
    builder.encodeStruct(controlMessages.RunResponseMessageParams,
                         runResponseMessageParams);
    responder.accept(builder.finish());
    return true;
  }

  function isControlMessage(message) {
    return message.getName() == controlMessages.kRunMessageId ||
           message.getName() == controlMessages.kRunOrClosePipeMessageId;
  }

  function ControlMessageHandler(interface_version) {
    this.interface_version = interface_version;
  }

  ControlMessageHandler.prototype.accept = function(message) {
    validateControlRequestWithoutResponse(message);
    return runOrClosePipe(message, this.interface_version);
  };

  ControlMessageHandler.prototype.acceptWithResponder = function(message,
      responder) {
    validateControlRequestWithResponse(message);
    return run(message, responder, this.interface_version);
  };

  var exports = {};
  exports.ControlMessageHandler = ControlMessageHandler;
  exports.isControlMessage = isControlMessage;

  return exports;
});
